metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/SyncService.scala

```diff
@@ -27,28 +39,57 @@ class SyncService[F[_]: Sync, N, A <: Agreement](
     blockStorage: BlockStorage[N, A],
     blockSyncPipe: BlockSyncPipe[F, A]#Right,
     getState: F[ProtocolState[A]],
-    fiberMap: FiberMap[F, A#PKey]
+    incomingFiberMap: FiberMap[F, A#PKey],
+    syncFiberMap: FiberMap[F, A#PKey],
+    rpcTracker: RPCTracker[F, SyncMessage[A]]
 )(implicit tracers: SyncTracers[F, A], storeRunner: KVStoreRunner[F, N]) {
+  import SyncMessage._
 
-  /** Request a block from a peer.
-    *
-    * Returns `None` if we're not connected or the request times out.
-    */
-  def getBlock(from: A#PKey, blockHash: A#Hash): F[Option[A#Block]] = ???
+  /** Request a block from a peer. */
+  private def getBlock(from: A#PKey, blockHash: A#Hash): F[Option[A#Block]] = {
+    for {
+      requestId <- RequestId[F]
+      request = GetBlockRequest(requestId, blockHash)
+      maybeResponse <- sendRequest(from, request)
+    } yield maybeResponse.map(_.block)
```

Comment:
Could you explain why using a `for` makes this referentially transparent in contrast to the previous version? (just curious)

Reply:
Sure, it was the change from `RequestId()` to `RequestId[F]`

Previously it was something like this:
```scala
def send(params): F[Response] = {
  val requestId = RequestId()
  val request = Request(requestId, params)
  Sync[F].delay(rpc(request))
}
```
So when we call `send` we always get a different request ID, so these two were different:
```scala
val a = send(1)
val b = a // Uses the same request ID
// vs
val a = send(1)
val b = send(1) // Uses a different request ID.
```

Now it's like this:
```scala
def send(params): F[Response] = 
  RequestId[F].map(Request(_, params)).flatMap(rpc)
```
When we call `send`, no side effect happens, so the two cases above would be equivalent.